src/fcntl.rs

```diff
@@ -596,39 +597,125 @@ pub fn fcntl(fd: RawFd, arg: FcntlArg) -> Result<c_int> {
+// `ManuallyDrop` is necessary to circumvent move out of `Drop` type error.
+#[cfg(not(any(target_os = "redox", target_os = "solaris")))]
+#[derive(Debug)]
+pub struct Flock<T: Flockable>(Option<T>);
+
+#[cfg(not(any(target_os = "redox", target_os = "solaris")))]
+impl<T: Flockable> Drop for Flock<T> {
+    fn drop(&mut self) {
+       if let Some(ref t) = self.0 {
+           // Result is ignored because flock has no documented failure cases.
+           _ = unsafe { libc::flock(t.as_raw_fd(), libc::LOCK_UN) };
+       }
+    }
+}
+
+#[cfg(not(any(target_os = "redox", target_os = "solaris")))]
+impl<T: Flockable> Deref for Flock<T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        if let Some(ref t) = self.0 {
+            t
+        } else { unreachable!() }
+    }
+}
+#[cfg(not(any(target_os = "redox", target_os = "solaris")))]
+impl<T: Flockable> DerefMut for Flock<T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        if let Some(ref mut t) = self.0 {
+            t
+        } else { unreachable!() }
+    }
+}
+
+#[cfg(not(any(target_os = "redox", target_os = "solaris")))]
+impl<T: Flockable> Flock<T> {
+    /// Obtain a/an flock.
+    ///
+    /// # Example
+    /// ```
+    /// # use std::fs::File;
+    /// # use nix::fcntl::{Flock, FlockArg};
+    /// fn do_stuff(file: File) -> nix::Result<()> {
+    ///     let lock = match Flock::lock(file, FlockArg::LockExclusive) {
+    ///         Ok(l) => l,
+    ///         Err((_, e)) => return Err(e),
+    ///     };
+    ///
+    ///     // Do stuff
+    ///
+    ///     Ok(())
+    /// } // File is unlocked once `lock` goes out of scope.
+    pub fn lock(t: T, args: FlockArg) -> std::result::Result<Self, (T, Errno)> {
+        let flags = match args {
+            FlockArg::LockShared => libc::LOCK_SH,
+            FlockArg::LockExclusive => libc::LOCK_EX,
+            FlockArg::LockSharedNonblock => libc::LOCK_SH | libc::LOCK_NB,
+            FlockArg::LockExclusiveNonblock => libc::LOCK_EX | libc::LOCK_NB,
+        };
+        match Errno::result(unsafe { libc::flock(t.as_raw_fd(), flags) }) {
+            Ok(_) => Ok(Self(Some(t))),
+            Err(errno) => Err((t, errno)),
         }
-    };
+    }
 
-    Errno::result(res).map(drop)
+    /// Remove the lock and return the object wrapped within.
+    ///
+    /// # Example
+    /// ```
+    /// # use std::fs::File;
+    /// # use nix::fcntl::{Flock, FlockArg};
+    /// fn do_stuff(file: File) -> nix::Result<()> {
+    ///     let lock = match Flock::lock(file, FlockArg::LockExclusive) {
+    ///         Ok(l) => l,
+    ///         Err((_,e)) => return Err(e),
+    ///     };
+    ///
+    ///     // Do critical section
+    ///
+    ///     // Unlock
+    ///     let file = lock.unlock();
+    ///
+    ///     // Do anything else
+    ///
+    ///     Ok(())
+    /// }
+    pub fn unlock(mut self) -> T {
+        _ = unsafe { libc::flock(self.0.as_ref().unwrap().as_raw_fd(), libc::LOCK_UN) };
```

Comment:
You should not ignore the error code.  There could still be, for example, an EBADF error if somebody implemented Flockable on a Clone type, contrary to our documentation.

Reply:
See your above comment on drop for full details, but I'll make the changes here.

Just want to note though that if someone ignored the documentation, and it causes some unexpected error, that's explicitly their problem because we marked the trait as `unsafe`. `EBADF` in particular would not be our problem imho, because they've broken our assumption that they have not cloned their fd.

Reply:
Yeah, it wouldn't be "our fault".  But ignoring an EBADF error could lead to some _very_ difficult bugs, such as when the user opens a new file with the same file descriptor and we close that new file on drop, resulting in some sort of race.